src/sundials/sundials_context.c

```diff
@@ -44,88 +62,105 @@ int SUNContext_Create(SUNComm comm, SUNContext* sunctx)
     comm = SUN_COMM_NULL;
   }
 
-#if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-  if (SUNProfiler_Create(comm, "SUNContext Default", &profiler)) return (-1);
-#endif
-
-#ifdef SUNDIALS_ADIAK_ENABLED 
-  adiak_init(&comm);
-  sunAdiakCollectMetadata();
-#endif
-
 #if SUNDIALS_LOGGING_LEVEL > 0 
 #if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_CreateFromEnv(comm, &logger))
-#else
-  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger))
-#endif
-  {
-    return (-1);
+  if (SUNLogger_CreateFromEnv(comm, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
   }
 #else
-#if SUNDIALS_MPI_ENABLED
-  if (SUNLogger_Create(comm, 0, &logger))
-#else
-  if (SUNLogger_Create(SUN_COMM_NULL, 0, &logger))
+  if (SUNLogger_CreateFromEnv(SUN_COMM_NULL, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
+  }
 #endif
-  {
-    return (-1);
+#else
+  if (SUNLogger_Create(NULL, 0, &logger)) {
+    return SUN_ERR_LOGGER_CORRUPT;
   }
-  SUNLogger_SetErrorFilename(logger, "");
-  SUNLogger_SetWarningFilename(logger, "");
-  SUNLogger_SetInfoFilename(logger, "");
-  SUNLogger_SetDebugFilename(logger, "");
+  SUNCheckCall(SUNLogger_SetErrorFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetWarningFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetInfoFilename(logger, ""));
+  SUNCheckCall(SUNLogger_SetDebugFilename(logger, ""));
 #endif
 
-  *sunctx = NULL;
-  *sunctx = (SUNContext)malloc(sizeof(struct _SUNContext));
-
-  if (*sunctx == NULL)
-  {
 #if defined(SUNDIALS_BUILD_WITH_PROFILING) && !defined(SUNDIALS_CALIPER_ENABLED)
-    SUNProfiler_Free(&profiler);
+  SUNCheckCall(SUNProfiler_Create(comm, "SUNContext Default", &profiler));
 #endif
-    SUNLogger_Destroy(&logger);
-    return (-1);
-  }
 
-  (*sunctx)->logger       = logger;
-  (*sunctx)->own_logger   = logger != NULL;
-  (*sunctx)->profiler     = profiler;
-  (*sunctx)->own_profiler = profiler != NULL;
+  sunctx->logger        = logger;
+  sunctx->own_logger    = logger != NULL;
+  sunctx->profiler      = profiler;
+  sunctx->own_profiler  = profiler != NULL;
+  sunctx->last_err      = 0;
+  sunctx->err_handler   = SUNErrHandler_Create(SUNLogErrHandlerFn, NULL);
+  sunctx->comm          = comm;
+
+  *sunctx_out = sunctx;
 
-  return (0);
+  return SUN_SUCCESS;
 }
 
-int SUNContext_GetProfiler(SUNContext sunctx, SUNProfiler* profiler)
+SUNErrCode SUNContext_GetLastError(SUNContext sunctx, SUNErrCode* last_err)
 {
-  if (sunctx == NULL)
+  *last_err = sunctx->last_err;
+  sunctx->last_err = SUN_SUCCESS;
+  return SUN_SUCCESS;
```

Comment:
A bit counterintuitive that the return value is not the last err. Could the out argument be the return parameter?

Reply:
We are trying to make (almost) all new functions return an error code and the value of interest as an out parameter. If you see inconsistencies in this for new functions in this PR, please point those out.

Reply:
I do think this function is a special case though, and it may be one of the exceptions where we return the value of interest, which also happens to be an error code, instead of an error code indicating success/failure of the function itself.